client/mm/libxc/binance.go

```diff
@@ -47,38 +47,180 @@ const (
+	return &binanceOrderBook{
+		book:                  newOrderBook(),
+		mktID:                 mktID,
+		updateQueue:           make(chan *binanceBookUpdate, 1024),
+		numSubscribers:        1,
+		baseConversionFactor:  baseConversionFactor,
+		quoteConversionFactor: quoteConversionFactor,
+		log:                   log,
+	}
 }
 
-func newBNCBook() *bncBook {
-	return &bncBook{
-		bids:           make([]*bookBin, 0),
-		asks:           make([]*bookBin, 0),
-		numSubscribers: 1,
+// convertBinanceBook converts bids an asks in the binance format,
+// with the conventional quantity and rate, to the DEX message format which
+// can be used to update the orderbook.
+func (b *binanceOrderBook) convertBinanceBook(binanceBids, binanceAsks [][2]json.Number) (bids, asks []*obEntry, err error) {
+	convert := func(updates [][2]json.Number) ([]*obEntry, error) {
+		convertedUpdates := make([]*obEntry, 0, len(updates))
+
+		for _, update := range updates {
+			price, err := update[0].Float64()
+			if err != nil {
+				return nil, fmt.Errorf("error parsing price: %v", err)
+			}
+
+			qty, err := update[1].Float64()
+			if err != nil {
+				return nil, fmt.Errorf("error parsing qty: %v", err)
+			}
+
+			convertedUpdates = append(convertedUpdates, &obEntry{
+				rate: calc.MessageRateAlt(price, b.baseConversionFactor, b.quoteConversionFactor),
+				qty:  uint64(qty * float64(b.baseConversionFactor)),
+			})
+		}
+
+		return convertedUpdates, nil
+	}
+
+	bids, err = convert(binanceBids)
+	if err != nil {
+		return nil, nil, err
 	}
+
+	asks, err = convert(binanceAsks)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return bids, asks, nil
 }
 
-type bncAssetConfig struct {
-	// symbol is the DEX asset symbol, always lower case
-	symbol string
-	// coin is the asset symbol on binance, always upper case.
-	// For a token like USDC, the coin field will be USDC, but
-	// symbol field will be usdc.eth.
-	coin string
-	// chain will be the same as coin for the base assets of
-	// a blockchain, but for tokens it will be the chain
-	// that the token is hosted such as "ETH".
-	chain            string
-	conversionFactor uint64
+// sync does an initial sync of the orderbook. When the first update is
+// received, it grabs a snapshot of the orderbook, and only processes updates
+// that come after the state of the snapshot. A goroutine is started that keeps
+// that keeps the orderbook in sync by repeating the sync process if an update
+// is ever missed.
+//
+// This function runs until the context is canceled. It must be started as
+// a new goroutine.
+func (b *binanceOrderBook) sync(ctx context.Context, getSnapshot func() (*binanceOrderbookSnapshot, error)) {
+	syncOrderbook := func(minUpdateID uint64) (uint64, bool) {
+		b.log.Debugf("Syncing %s orderbook. First update ID: %d", b.mktID, minUpdateID)
+
+		const maxTries = 5
+		for i := 0; i < maxTries; i++ {
+			if i > 0 {
+				time.Sleep(2 * time.Second)
+			}
```

Comment:
No specific reason.. this will only be called if there's a bug on the Binance side where they sent an order book update, then in a snapshot response they sent an earlier state than the update that they already sent. It just gives Binance two seconds to recover from their weird state.

Reply:
Oh, okay.